### ISSUE TITLE
Intent: implicit functions allowed in function names

### ISSUE THREAD

**Issue by NSoiffer** [w3c/mathml#380]
As with my other comments on `intent`'s grammar, this is mainly just to make sure people are aware of this, not that I'm suggesting a change...

Currently the intent grammar allows only `NCName` or `argref` for the name of a function. That means `f(x)(y)` is not legal. However, because of `argref`, you can still get that:
```
<mrow intent='$f-of-x($y)'>
   <mrow arg='f-of'x' intent='$f($x)'>
     <mi arg='f'>f</mi>
     <mo>(</mo>
     <mi arg='x'>x</mi>
     <mo>)</mo>
   <mrow>
   <mi arg='y'>y</mi>
</mrow>
```

This has the same affect as `intent='$f($x)($y)'` but is more indirect/obscure in what is happening. Good or bad?

**Comment by brucemiller** [comment]
I had original written
```
intent   ::= literal | selector | intent '(' intent [ ',' intent ]* ')'
```
which allows `intent="f(x)(y)"`. I hadn't noticed that @davidcarlisle's rewrite removed that capability. I don't know if that was intentional?

As @NSoiffer says, you can get around it by introducing extra `mrow`s; but by doing that you barely need intent anyway :>

**Comment by davidcarlisle** [comment]
Yes this one was intentional, if you allow `f(x)(y)` as intent you'll need a meta-intent to say if that is function applicaion or invisible times between f(x) and (y) (well actually not, as intent doesn't have infix multiply, but it still looked weird to me)

 an alternative to adding extra argref would be to use a function application literal so `apply( f(x), y  )`  which probably leads to a better reading in any case.

as @brucemiller just indicated while I was writing this, it would be easy to allow this, just change the grammar to allow `intent` before the `(` but...

**Comment by dginev** [comment]
My follow-up minimalist grammar proposal removed the capability, since adding it in allows for two different ways of expressing the same feature.

Neil's example of arg-based nesting is possible in either case, so it ought to remain, if we wanted a single way of expressing the chaining, in a minimalist view.

To restate my perspective from #252, I think the smaller we manage to make the final intent language, the easier it would be to get third-party implementations afterwards.

Edit: David correctly pointed out it is not the grammar that should be made smaller, but the language induced by it.

**Comment by davidcarlisle** [comment]
I should have said, while not allowing it in this reduced grammar was intentional, I don't necessarily object to allowing it. As with allowing an empty argument list there is an argument that says we should not disallow things that might possibly be useful, even if  we feel using them may not be best practice. If we allow it, it allows community to come to consensus on whether or not it is useful  in practice.

**Comment by davidcarlisle** [comment]
@dginev there are arguments either way but I don't think "the grammar is smaller or simpler" is an argument for either form, It's acually smaller if you replace `(NCName | argref)` by `intent` .

**Comment by brucemiller** [comment]
It is certainly not simpler for the annotator if they have to add `mrow`s to the markup. I understood that an important usecase was to be able to disambiguate existing markup where you cannot or should not modify the markup other than by annotation.  (as well as creating the markup from scratch, of course).

**Comment by NSoiffer** [comment]
Two potential use cases for complicated function "names" leap to mind:
1. D^2 f
2. (f ∘g)x

The first case might be represented as `derivative($sup, $f)` although that won't speak well -- you probably want to hear something like "D squared of f". Perhaps a more natural (and better) solution is to write `power(D, $sup)($f)`. That's an example of a complicated head. Of course one could also write `D-squared($f)`. So I don't know if this is convincing.

The second case has the natural(?) representation as `compose($f, $g)($x)`. I don't think @davidcarlisle's suggestion of using `apply` in these cases (e.g., `apply(compose($f, $g), $x)`) makes sense unless we say `apply` is core and AT should probably speak it as "of" as opposed to "apply ... to ...".

**Comment by brucemiller** [comment]
I like the revised grammar in #376

**Comment by davidcarlisle** [comment]
addressed at a2a141d
